### PR TITLE
Improve on-heap aggregator footprint estimates.

### PR DESCRIPTION
**Pull request by gianm**
Add a "guessAggregatorHeapFootprint" method to AggregatorFactory that mitigates #6743 by enabling heap footprint estimates based on a specific number of rows. The idea is that at ingestion time, the number of rows that go into an aggregator will be 1 (if rollup is off) or will likely be a small number (if rollup is on).

It's a heuristic, because of course nothing guarantees that the rollup ratio is a small number. But it's a common case, and I expect this logic to go wrong much less often than the current logic. Also, when it does go wrong, users can fix it by lowering maxRowsInMemory or maxBytesInMemory. The current situation is unintuitive: when the estimation goes wrong, users get an OOME, but actually they need to *raise* these limits to fix it.

I don't think this is an ideal solution. In the future, I'd like to see something that involves the IncrementalIndex having much more direct control over how much memory it uses. Probably in concert with implementing off-heap resizeable aggregators, which is a feature that would be very nice to have for other reasons too. But in the meantime, the approach in this patch is simple, non-invasive, and I think it will help people.

### CODE REVIEW

File: extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/SketchAggregatorFactory.java [hunk 1]
```diff
@@ -170,6 +171,25 @@ public int getSize()
...
+  @Override
+  public int guessAggregatorHeapFootprint(long rows)
+  {
+    final int maxEntries = size * 2;
+    final int expectedEntries;
+
+    if (rows > maxEntries) {
+      expectedEntries = maxEntries;
+    } else {
+      final int minEntries = 1 << Util.MIN_LG_ARR_LONGS;
```
Comment (kfaraz) [inline extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/SketchAggregatorFactory.java:183]: Should `minEntries = 1 << Util.MIN_LG_ARR_LONGS` and the largest possible preamble, `Family.UNION.getMaxPreLongs() << 3` be taken out into constants?

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java [hunk 2]
```diff
@@ -114,10 +122,13 @@
...
     long maxAggregatorIntermediateSize = ((long) Integer.BYTES) * incrementalIndexSchema.getMetrics().length;
-    maxAggregatorIntermediateSize += Arrays.stream(incrementalIndexSchema.getMetrics())
-                                           .mapToLong(aggregator -> aggregator.getMaxIntermediateSizeWithNulls()
-                                                                    + Long.BYTES * 2L)
-                                           .sum();
+
+    for (final AggregatorFactory aggregator : incrementalIndexSchema.getMetrics()) {
+      final long rows =
+          incrementalIndexSchema.isRollup() ? ROLLUP_RATIO_FOR_AGGREGATOR_FOOTPRINT_ESTIMATION : 1;
+      maxAggregatorIntermediateSize += (long) aggregator.guessAggregatorHeapFootprint(rows) + 2 * Long.BYTES;
```
Comment (kfaraz) [inline processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:129]: Nit: Is the term `2 * Long.BYTES (= 16)` needed now?
It seems that that the result of the guess would be big enough even when `rows = 1` to make this term relatively much less significant.

Comment (gianm) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java:129]: It's still useful for aggregators that have a small data size, like longSum etc. For those, the overheads are bigger than the intermediate data size.

**Comment by gianm** [comment]
@kfaraz thanks for the review. I pushed up some of the changes you suggested.

File: processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java (hunk 2 above)
Comment (kfaraz) [reply processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java]: Makes sense 👍

**Comment by kfaraz** [comment]
Thanks for the changes, @gianm .

On running ingestion of sample "wikipedia" data, I got the following persisted stats in the indexing logs.

Before the changes 
(total 38 persistence cycles, 652 rows persisted in each cycle):
```
...
2021-11-24T15:55:26,139 INFO [[index_parallel_test_sketch_lolhhmjb_2021-11-24T15:54:54.532Z]-appenderator-persist] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Persisted stats: processed rows: [1304], persisted rows[652], sinks: [1], total fireHydrants (across sinks): [1], persisted fireHydrants (across sinks): [1]
2021-11-24T15:55:26,244 INFO [[index_parallel_test_sketch_lolhhmjb_2021-11-24T15:54:54.532Z]-appenderator-persist] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Persisted stats: processed rows: [1956], persisted rows[652], sinks: [1], total fireHydrants (across sinks): [2], persisted fireHydrants (across sinks): [2]
2021-11-24T15:55:26,353 INFO [[index_parallel_test_sketch_lolhhmjb_2021-11-24T15:54:54.532Z]-appenderator-persist] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Persisted stats: processed rows: [2608], persisted rows[652], sinks: [1], total fireHydrants (across sinks): [3], persisted fireHydrants (across sinks): [2]
2021-11-24T15:55:26,436 INFO [[index_parallel_test_sketch_lolhhmjb_2021-11-24T15:54:54.532Z]-appenderator-persist] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Persisted stats: processed rows: [3260], persisted rows[652], sinks: [1], total fireHydrants (across sinks): [4], persisted fireHydrants (across sinks): [2]
2021-11-24T15:55:26,509 INFO [[index_parallel_test_sketch_lolhhmjb_2021-11-24T15:54:54.532Z]-appenderator-persist] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Persisted stats: processed rows: [3912], persisted rows[652], sinks: [1], total fireHydrants (across sinks): [5], persisted fireHydrants (across sinks): [2]
2021-11-24T15:55:26,578 INFO [[index_parallel_test_sketch_lolhhmjb_2021-11-24T15:54:54.532Z]-appenderator-persist]
...
```

After the changes 
(total 1 persistence cycle, ~24k rows persisted in one cycle):
```
2021-11-25T03:16:30,200 INFO [[index_parallel_test_sketchy_ghobnhcj_2021-11-25T03:15:57.793Z]-appenderator-persist] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Persisted stats: processed rows: [24433], persisted rows[24433], sinks: [1], total fireHydrants (across sinks): [1], persisted fireHydrants (across sinks): [1]
```

**Review by kfaraz — APPROVED** [review]
LGTM

**Review by asdf2014 — APPROVED** [review]
LGTM 👍

**Pull request merged**